### PR TITLE
Implement Recipes

### PR DESCRIPTION
**Pull request by Maxqia**
[API](https://github.com/SpongePowered/SpongeAPI/pull/1419) | **Common**

### CODE REVIEW

File: src/main/java/org/spongepowered/common/mixin/core/item/recipe/MixinFurnaceRecipies.java [hunk 1]
```diff
@@ -0,0 +1,47 @@
+package org.spongepowered.common.mixin.core.item.recipe;
```
Comment (kashike) [inline src/main/java/org/spongepowered/common/mixin/core/item/recipe/MixinFurnaceRecipies.java]: Missing licence header.

File: src/main/java/org/spongepowered/common/mixin/core/item/recipe/MixinFurnaceRecipies.java [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+import java.util.Map;
```
Comment (kashike) [inline src/main/java/org/spongepowered/common/mixin/core/item/recipe/MixinFurnaceRecipies.java]: Wrong import order.

File: src/main/java/org/spongepowered/common/mixin/core/item/recipe/MixinFurnaceRecipies.java [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+import org.spongepowered.api.item.recipe.SmeltingRegistry;
+import org.spongepowered.asm.mixin.Mixin;
+import org.spongepowered.asm.mixin.Shadow;
+import org.spongepowered.common.item.inventory.util.ItemStackUtil;
+
+import net.minecraft.item.ItemStack;
```
Comment (kashike) [inline src/main/java/org/spongepowered/common/mixin/core/item/recipe/MixinFurnaceRecipies.java]: Wrong import order.

File: src/main/java/org/spongepowered/common/mixin/core/item/recipe/MixinFurnaceRecipies.java [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+    @Shadow public abstract void addSmeltingRecipe(ItemStack input, ItemStack stack, float experience);
+
+    @Override
+    public void addSmeltingRecipe(org.spongepowered.api.item.inventory.ItemStack input, org.spongepowered.api.item.inventory.ItemStack output, float exp) {
+        addSmeltingRecipe(ItemStackUtil.toNative(input), ItemStackUtil.toNative(output), exp);
+    }
+
+    @Override
+    public boolean removeSmeltingRecipe(org.spongepowered.api.item.inventory.ItemStack input) {
+        ItemStack stack = smeltingList.remove(input);
```
Comment (kashike) [inline src/main/java/org/spongepowered/common/mixin/core/item/recipe/MixinFurnaceRecipies.java]: `this.`

File: src/main/java/org/spongepowered/common/mixin/core/item/recipe/MixinFurnaceRecipies.java [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+                return true;
+            }
+            throw new RuntimeException("It was in one map but not the other?");
+        }
+        return false;
+    }
+
+    @Override
+    public org.spongepowered.api.item.inventory.ItemStack getMapping(org.spongepowered.api.item.inventory.ItemStack stack) {
+        return ItemStackUtil.cloneDefensive(smeltingList.get(stack));
```
Comment (kashike) [inline src/main/java/org/spongepowered/common/mixin/core/item/recipe/MixinFurnaceRecipies.java]: `this.`

File: src/main/java/org/spongepowered/common/mixin/core/item/recipe/MixinFurnaceRecipies.java [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+    }
+
+    @Override
+    public float getExperience(org.spongepowered.api.item.inventory.ItemStack stack) {
+        return experienceList.get(stack);
```
Comment (kashike) [inline src/main/java/org/spongepowered/common/mixin/core/item/recipe/MixinFurnaceRecipies.java]: `this.`

**Review by kashike — CHANGES_REQUESTED** [review]

File: src/main/java/org/spongepowered/common/mixin/core/item/recipe/MixinFurnaceRecipies.java [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+    }
+
+    @Override
+    public boolean removeSmeltingRecipe(org.spongepowered.api.item.inventory.ItemStack input) {
+        ItemStack stack = smeltingList.remove(input);
+        if (stack != null) {
+            if (experienceList.remove(stack) != null) {
+                return true;
+            }
+            throw new RuntimeException("It was in one map but not the other?");
```
Comment (liach) [inline src/main/java/org/spongepowered/common/mixin/core/item/recipe/MixinFurnaceRecipies.java]:
``` java
throw new AssertionError("It was on one map but not the other.");
```

File: src/main/java/org/spongepowered/common/mixin/core/item/recipe/MixinFurnaceRecipies.java [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+@Mixin(FurnaceRecipes.class)
+public abstract class MixinFurnaceRecipies implements SmeltingRegistry {
+
+    @Shadow private final Map<ItemStack, ItemStack> smeltingList = null;
+    @Shadow private final Map<ItemStack, Float> experienceList = null;
+    @Shadow public abstract void addSmeltingRecipe(ItemStack input, ItemStack stack, float experience);
+
+    @Override
+    public void addSmeltingRecipe(org.spongepowered.api.item.inventory.ItemStack input, org.spongepowered.api.item.inventory.ItemStack output, float exp) {
+        addSmeltingRecipe(ItemStackUtil.toNative(input), ItemStackUtil.toNative(output), exp);
```
Comment (liach) [inline src/main/java/org/spongepowered/common/mixin/core/item/recipe/MixinFurnaceRecipies.java]: Is `toNative` simply for casting?

**Comment by liach** [comment]
Why do you never update submodule?

File: src/main/java/org/spongepowered/common/mixin/core/item/recipe/MixinFurnaceRecipies.java (hunk 8 above)
Comment (kashike) [reply src/main/java/org/spongepowered/common/mixin/core/item/recipe/MixinFurnaceRecipies.java]: Yes.

**Comment by Maxqia** [comment]
Now with more testing!
EDIT: removed itemstackutil changes

File: src/main/java/org/spongepowered/common/mixin/core/item/recipe/MixinFurnaceRecipes.java [hunk 9]
```diff
@@ -79,8 +80,8 @@ private ItemStack getInternalMapping(org.spongepowered.api.item.inventory.ItemSt
     }
 
     @Override
-    public float getExperience(org.spongepowered.api.item.inventory.ItemStack stack) {
-        return this.experienceList.get(stack);
+    public Optional<Float> getExperience(org.spongepowered.api.item.inventory.ItemStack stack) {
```
Comment (liach) [inline src/main/java/org/spongepowered/common/mixin/core/item/recipe/MixinFurnaceRecipes.java]: There should be an `OptionalDouble` available in `java.util`.

Comment (Maxqia) [reply src/main/java/org/spongepowered/common/mixin/core/item/recipe/MixinFurnaceRecipes.java]: That exists?!? ..., but it's kind of stupid, it always says it's present even if it's null

File: src/main/java/org/spongepowered/common/item/recipe/SpongeSmeltingRecipe.java [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+        this.experience = exp;
+        return this;
+    }
+
+    @Override
+    public SmeltingRecipe build() {
+        checkNotNull(ingredient, "ingredient");
+        checkNotNull(result, "result");
+        checkNotNull(experience, "experience");
+        return this;
```
Comment (Limeth) [inline src/main/java/org/spongepowered/common/item/recipe/SpongeSmeltingRecipe.java]: I am confused. This is not how builders are supposed to work.

File: src/main/java/org/spongepowered/common/item/recipe/crafting/SpongeShapedCraftingRecipeBuilder.java [hunk 11]
```diff
@@ -0,0 +1,153 @@
...
+
+    public static <T> Map<T, ItemStack> mapItemStack(ImmutableMap<T, ItemStackSnapshot> map) {
+        HashMap<T, ItemStack> returnVal = new HashMap<>();
+        for (Entry<T,ItemStackSnapshot> entry : map.entrySet()) {
+            returnVal.put(entry.getKey(), entry.getValue().createStack());
+        }
+        return returnVal;
+    }
+
+    public static <T> ImmutableMap<T, ItemStackSnapshot> mapItemStackSnapshot(Map<T, ItemStack> map) {
```
Comment (liach) [inline src/main/java/org/spongepowered/common/item/recipe/crafting/SpongeShapedCraftingRecipeBuilder.java:145]: These should move to util classes.

File: src/main/java/org/spongepowered/common/mixin/core/item/recipe/MixinCraftingManager.java [hunk 12]
```diff
@@ -0,0 +1,129 @@
...
+        ((IMixinShapedRecipes) recipe).setShape(this.shapedAisle);
+        ((IMixinShapedRecipes) recipe).setIngredientMap(this.shapedIngredients);
+        this.shapedAisle = null;
+        this.shapedIngredients = null;
+        return recipe;
+    }
+
+    @Override
+    public Optional<ImmutableCollection<ItemStackSnapshot>> getResults(GridInventory grid) {
+        throw new NotImplementedException("TODO"); // dunno how to convert a gridinventory to a IInventory...
```
Comment (liach) [inline src/main/java/org/spongepowered/common/mixin/core/item/recipe/MixinCraftingManager.java:126]: :trollface:

File: src/main/java/org/spongepowered/common/mixin/core/item/recipe/MixinIRecipe.java [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+import org.spongepowered.asm.mixin.Shadow;
+import org.spongepowered.common.SpongeImpl;
+
+@Mixin(IRecipe.class)
+public interface MixinIRecipe extends CraftingRecipe {
+
+    @Shadow net.minecraft.item.ItemStack getRecipeOutput();
+
+    @Override
+    default ImmutableCollection<ItemStackSnapshot> getResults() {
```
Comment (liach) [inline src/main/java/org/spongepowered/common/mixin/core/item/recipe/MixinIRecipe.java:44]: Why are you using `ImmutableCollection`?

**Comment by kashike** [comment]
I'm going to close this in favour of #1395.

**Pull request closed without merging**